---
title: "R Notebook"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
---



```{r, message=FALSE, warning=FALSE}
library("knitr")
opts_chunk$set(cache = TRUE, message = FALSE, warning = FALSE)

library(tidyverse)
library(data.table)
library(useful)
library(biomaRt)
library(ggrepel)
library(ggthemes)
library(magrittr)
library(DESeq2)
library(cowplot)
library(pheatmap)
library(broom)
library(openxlsx)
```

# QC

```{r, eval=FALSE, include=FALSE}
datadir <- "../data/get.broadinstitute.org/pkgs/SN0137582/"
fname <- list.files(datadir, pattern = "alignment_summary_metrics")[1]
files.qc <- list.files(datadir, pattern = "alignment_summary_metrics")
qc.df <- do.call(rbind, lapply(files.qc, function(fname) {
  fname2 <- gsub("alignment_summary", "rna", fname)
  read.table(sprintf("%s/%s", datadir, fname), sep = "\t", header = T) %>%
    dplyr::filter(CATEGORY == "PAIR") %>%
    dplyr::select(TOTAL_READS, PCT_PF_READS_ALIGNED) %>%
    cbind(
      read.table(sprintf("%s/%s", datadir, fname2), sep = "\t", header = T, nrows = 1) %>%
        dplyr::select(PCT_INTRONIC_BASES, PCT_INTERGENIC_BASES, PCT_RIBOSOMAL_BASES, PCT_CODING_BASES, PCT_UTR_BASES)
    )
}))

qc.df$Position <- gsub("^(\\d)_.*?_(.\\d\\d).*", "\\2", files.qc)
qc.df$rep <- gsub("^(\\d)_.*?_(.\\d\\d).*", "\\1", files.qc)

## library complexity/duplicates
qc.df %<>%
  dplyr::inner_join(fread("../data/qc_duplicates.csv")[, .(PERCENT_DUPLICATION, ESTIMATED_LIBRARY_SIZE, Position, rep = as.character(rep))])

qc.df <- data.table(qc.df %>% dplyr::select(-rep))[, lapply(.SD, mean), by = Position]
head(qc.df)

saveRDS(qc.df, "../data/qc_df.rds")
```

```{r}
qc.df <- readRDS("../data/qc_df.rds")
```




# Init data




## BiomaRt

```{r, eval=FALSE, include=FALSE}
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
listAttributes(ensembl) %>% subset(grepl("entr", name, ignore.case = T))
ensembl.map <- getBM(mart = ensembl, attributes = c("ensembl_gene_id", "ensembl_transcript_id", "hgnc_symbol", "entrezgene")) %>%
  subset(!is.na(hgnc_symbol))
saveRDS(ensembl.map, "..data/biomart_enmsebl_map.rds")
```

```{r}
ensembl.map <- readRDS("../data/biomart_enmsebl_map.rds")
```


## metadata

D12 was an outlier by QC standards (low read depth, low alignment to genome)

```{r, message=FALSE, warning=FALSE}
metadata <- read.xlsx("../data/Raychaudhuri_PlateMap_Final.xlsx") %>%
  tidyr::separate(Sample.Identifier, c("donor", "cell"), sep = ":") %>% 
  dplyr::select(Position, donor, cell) %>%
  dplyr::left_join(
    rbind(
      data.frame(donor = unlist(strsplit("10040772,10038559,10043467,10043449,28778769,10030474", split = ",")), status = "oa", stringsAsFactors = F),
      data.frame(donor = unlist(strsplit("10029276,10024302,10018377,10029272,10024301,10008438,10024253", split = ",")), status = "ra", stringsAsFactors = F)
    )
  )
row.names(metadata) <- metadata$Position

dds.coldata <- metadata#[samples.use, , drop = F]
dds.coldata$cell <- gsub("\\+", "p", dds.coldata$cell)
dds.coldata$cell <- gsub("-", "n", dds.coldata$cell)
```

## expression

```{r, eval=FALSE, include=FALSE}
data.dir <- "../data/Kallisto/"
exprs.kallisto <- do.call(rbind, list.files(data.dir, recursive = T, pattern = "abundance.tsv", full.names = T) %>% 
  lapply(function(fname) {
    fread(fname)[, .(Position = substr(fname, 62, 64), ensembl_transcript_id = gsub("\\..*", "", target_id), tpm, est_counts)]
  }))

## collapse to gene level
exprs.kallisto <- data.table(exprs.kallisto, key = "ensembl_transcript_id")[
    data.table(ensembl.map, key = "ensembl_transcript_id"), nomatch = 0
  ][!is.na(hgnc_symbol) & hgnc_symbol != ""
    , .(tpm = sum(tpm), est_counts = sum(est_counts)), by = .(hgnc_symbol, Position)
  ]

saveRDS(exprs.kallisto, "../data/exprs_kallisto.rds")
```

```{r}
exprs.kallisto <- readRDS("../data/exprs_kallisto.rds")
```



# Filter 

## Samples

Remove empty cells and outliers (D12). 

```{r}
samples.keep <- subset(metadata, Position != "D12" & cell != "Empty")$Position
exprs.kallisto %<>% dplyr::filter(Position %in% samples.keep)
# exprs.rsem %<>% dplyr::filter(Position %in% samples.keep)

# length(unique(exprs.kallisto$Position)); length(unique(exprs.rsem$Position))
```


## Genes

At least X counts in at least N samples

```{r}
genes.expressed.kallisto <- data.table(exprs.kallisto)[est_counts > 5, .N, by = hgnc_symbol][N > 10, hgnc_symbol]
# genes.expressed.rsem <- data.table(exprs.rsem)[expected_count > 5, .N, by = hgnc_symbol][N > 10, hgnc_symbol]
# genes.expressed <- intersect(genes.expressed.kallisto, genes.expressed.rsem)
# length(genes.expressed.kallisto); length(genes.expressed.rsem); length(genes.expressed)

length(genes.expressed.kallisto); length(unique(exprs.kallisto$hgnc_symbol))
exprs.kallisto.filtered <- dplyr::filter(exprs.kallisto, hgnc_symbol %in% genes.expressed.kallisto)
# exprs.rsem %<>% dplyr::filter(hgnc_symbol %in% genes.expressed)
```

# Normalize 


<!-- ```{r kallisto-tpm} -->
<!-- exprs.kallisto.tpm <- exprs.kallisto %>%  -->
<!--   dplyr::select(Position, hgnc_symbol, tpm) %>% -->
<!--   tidyr::spread(Position, tpm) %>%  -->
<!--   data.frame -->
<!-- row.names(exprs.kallisto.tpm) <- exprs.kallisto.tpm$hgnc_symbol -->
<!-- exprs.kallisto.tpm$hgnc_symbol <- NULL -->

<!-- exprs.kallisto.tpm <- log2(1 + exprs.kallisto.tpm) -->
<!-- ``` -->

```{r kallisto-counts}

exprs.kallisto.counts <- exprs.kallisto.filtered %>%
  dplyr::select(Position, hgnc_symbol, est_counts) %>%
  tidyr::spread(Position, est_counts) %>% 
  data.frame

row.names(exprs.kallisto.counts) <- exprs.kallisto.counts$hgnc_symbol
exprs.kallisto.counts$hgnc_symbol <- NULL

dds.kallisto <- DESeqDataSetFromMatrix(countData = round(exprs.kallisto.counts),
                              colData = dds.coldata[colnames(exprs.kallisto.counts), ],
                              design = ~cell)
dds.kallisto <- DESeq2::estimateSizeFactors(dds.kallisto)
dds.kallisto <- DESeq2::estimateDispersions(dds.kallisto)
exprs.kallisto.counts <- DESeq2::getVarianceStabilizedData(dds.kallisto)

# exprs.kallisto.counts <- log2(1 + exprs.kallisto.counts)

```


# PCA

A few takeaways from the plots below: 
- Kallisto Counts does not work well with DEseq VST normalization
- RSEM Counts separates TCM from TRreg, both TPM plots do not
- DR+/27+ mixes with DR-/27-

```{r}
pca.res.kallisto.counts <- prcomp(exprs.kallisto.counts)$rotation[, 1:2] %>% 
  data.frame %>% tibble::rownames_to_column("Position") %>% 
  dplyr::mutate(type = "kallisto.counts")
# pca.res.kallisto.tpm <- prcomp( exprs.kallisto.tpm)$rotation[, 1:2] %>% 
#   data.frame %>% tibble::rownames_to_column("Position") %>% 
#   dplyr::mutate(type = "kallisto.tpm")
# pca.res.rsem.counts <- prcomp(exprs.rsem.counts)$rotation[, 1:2] %>% 
#   data.frame %>% tibble::rownames_to_column("Position") %>% 
#   dplyr::mutate(type = "rsem.counts")
# pca.res.rsem.tpm <- prcomp(exprs.rsem.tpm)$rotation[, 1:2] %>% 
#   data.frame %>% tibble::rownames_to_column("Position") %>% 
#   dplyr::mutate(type = "rsem.tpm")
# 
# pca.res.all <- do.call(rbind, list(pca.res.rsem.tpm, pca.res.kallisto.tpm, 
#                                    pca.res.rsem.counts, pca.res.kallisto.counts))
```

## PC2 varies with cell type

```{r}
# pca.res.all %>%
#   dplyr::inner_join(metadata) %>%
#   ggplot(aes(PC1, PC2, col = cell)) + geom_point(size = 3, alpha = .7) +
#     scale_color_brewer(type = "qual") +
#     geom_rangeframe() + theme_tufte() +
#     theme(plot.title = element_text(hjust = 0.5, size = 16)) + 
#     facet_wrap(~type, scales = "free")

pca.res.kallisto.counts %>%
  dplyr::inner_join(metadata) %>%
  ggplot(aes(PC1, PC2, col = cell)) + geom_point(size = 3, alpha = .7) +
    scale_color_brewer(type = "qual") +
    geom_rangeframe() + theme_tufte() +
    theme(plot.title = element_text(hjust = 0.5, size = 16)) 

```

For downstream analyses, define the groups as ordered by PC2. Note that for TCM/Treg and DR+27+/DR-27-, the ordering is partial. 

```{r}
cell.levels <- c("Tnaive", "Treg", "TCM", "DR-27+", "DR+27+", "DR-27-", "DR+27-")
```


## PC gene loadings

```{r}
loadings.pc2 <- prcomp(t(exprs.kallisto.counts))$rotation[, 2]
genes.pc2 <- c(head(loadings.pc2[order(loadings.pc2)], 20), tail(loadings.pc2[order(loadings.pc2)], 20))

anno.col.df <- metadata %>% 
  dplyr::filter(Position %in% colnames(exprs.kallisto.counts)) %>% 
  dplyr::mutate(cell = factor(cell, levels = cell.levels)) %>% 
  dplyr::arrange(cell)
row.names(anno.col.df) <- anno.col.df$Position

exprs.kallisto.counts[names(genes.pc2), anno.col.df$Position] %>% 
  pheatmap(labels_col = rep("", 90), 
           annotation_col = anno.col.df[, "cell", drop = F], 
           annotation_row = data.frame(loading = genes.pc2),
           cluster_cols = F, cluster_rows = T, scale = "row")

```


## What does PC1 explain? 

*read depth?* no (r=-.06, p=.56)
*library complexity?* (i.e. duplicate rate)
*donor?* some, FDR < 0.01 for 10024302, 10043467
*well column?* almost no (q=0.008 for one column)
*well row?* no
*cell type?* not from linear model, but maybe to distinguish 


```{r}
x <- pca.res.kallisto.counts %>% 
  # dplyr::filter(type == "kallisto.counts") %>% 
  dplyr::inner_join(qc.df) %>%
  dplyr::inner_join(metadata)

tmp <- dplyr::select(x, Position, PC1, TOTAL_READS:ESTIMATED_LIBRARY_SIZE) %>% 
  tidyr::gather(key, val, -Position, -PC1)
data.table(tmp)[, tidy(cor.test(~ val + PC1, .SD)), by = key][
  , q.value := p.adjust(p.value)
][order(q.value), .(key, estimate, sprintf("%.8f", q.value))]

data.table(tmp)[, tidy(cor.test(~ val + PC1, .SD)), by = key][
  , q.value := p.adjust(p.value)
][order(q.value)] %>% with(as.integer(-log10(q.value)))

```

As a sanity check, visualize one of these relationships: 

```{r}
x %>% ggplot(aes(PC1, PCT_PF_READS_ALIGNED)) + geom_point() + geom_smooth(method = lm, se = F)
```


# Differential Expression

Limma, use disease status as covariate. 
Highlight expression of markers across groups (like I did for Deepak)


*todo* Effect of including RA/OA status as covariate (plot coefficients against each other)

## LIMMA

```{r, include=FALSE}
library(openxlsx)
library(limma)

groups.to.compare <- setdiff(unique(as.character(metadata$cell)), c("DR+27-", "Empty")) %>% as.list
names(groups.to.compare) <- unlist(groups.to.compare)
groups.to.compare$all <- setdiff(unique(as.character(metadata$cell)), c("DR+27-", "Empty"))
groups.to.compare$memory <- setdiff(unique(as.character(metadata$cell)), c("DR+27-", "Empty", "Naive"))
groups.to.compare$memoryEffector <- c("DR+27+", "DR-27-", "DR-27+")
names(groups.to.compare) <- gsub("\\+", "p", names(groups.to.compare))
names(groups.to.compare) <- gsub("-", "n", names(groups.to.compare))

wb.deseq <- createWorkbook()
limma.res.all <- data.frame()
for (other.group in names(groups.to.compare)) {
  print(other.group)
  addWorksheet(wb.deseq, other.group)
  
  samples.use <- metadata %>% 
    subset(cell %in% c("DR+27-", unlist(groups.to.compare[other.group]))) %>%
    rownames %>%
    intersect(colnames(exprs.kallisto.counts))
  
  design.mat <- model.matrix(~(cell == "DR+27-"), metadata[samples.use, ])
  limma.res <- lmFit(exprs.kallisto.counts[, samples.use], design = design.mat) %>% 
    eBayes() %>% 
    topTable(number = 5e4) %>% 
    tibble::rownames_to_column("feature") 
    # dplyr::select(-matches("donor")) %>%
    # dplyr::rename(logFC = cell.....DR.27..TRUE)

  limma.res.all %<>% rbind(limma.res %>% dplyr::mutate(other.group))
  
  writeDataTable(wb.deseq, other.group, limma.res)
  saveWorkbook(wb.deseq, file = "../results/DiffExp.xlsx", overwrite = TRUE)
}

```


```{r}
data.table(limma.res.all)[
  , direction := ifelse(logFC > 0, "up", "down")
][
  abs(logFC) > 1 & adj.P.Val < .01
][
  , .N, by = .(direction, other.group)
] %>% 
  ggplot(aes(reorder(other.group, N), N, fill = direction)) + 
    geom_bar(stat = "identity", position = "dodge") + 
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
    labs(title = "Number of Diff Exp Genes\n(FDR < .01, |logFC| > 1)", 
         x = "Comparison Group") 
```

## DESeq2



```{r}

```








# Markers

For consistency, we will order the cell types by PC2. 

## Pre-specified

We replaced CRTH2 with PTGDR2, and PRF with PRF1

```{r}
markers <- list()
markers$`transcription factors` <- unlist(strsplit("TBX21,RORC,GATA3,FOXP3,IKZF2,BCL6,PRDM1,MAF", split = ","))
markers$`cytokines` <- unlist(strsplit("IFNG,IL4,IL5,IL13,IL17A,IL17F,TNF,IL2,IL10", split = ","))
markers$`killer molecules` <- unlist(strsplit("GZMA,GZMB,PRF1,GZMK,GNLY,NKG7", split = ","))
markers$`chemokine receptors` <- unlist(strsplit("CXCR3,CCR4,CCR6,CCR2,PTGDR2", split = ","))

```


```{r}
plot.markers <- function(exprs.df, markers.use) {
  markers.use %<>% intersect(row.names(exprs.df))
  
  exprs.df <- exprs.df[markers.use, ] %>%
    data.frame %>%
    tibble::rownames_to_column("symbol") %>%
    tidyr::gather(Position, value, -symbol) %>% 
    dplyr::inner_join(metadata) %>%
    dplyr::mutate(cell = factor(cell, levels = cell.levels))
  
  exprs.df %>% 
    dplyr::filter(cell != "Empty") %>%
    ggplot(aes(cell, value, fill = cell == "DR+27-")) + 
      geom_boxplot(position = position_dodge(1)) +
      facet_wrap(~symbol, scales = "free_y") + 
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
      scale_fill_manual(values = c("white", "red")) + 
      guides(fill = FALSE, col = FALSE) + 
      labs(y = "log2 counts", x = "cell type")
}

plot.markers(exprs.kallisto.counts, markers$cytokines)
plot.markers(exprs.kallisto.counts, markers$`transcription factors`)
plot.markers(exprs.kallisto.counts, markers$`killer molecules`)
plot.markers(exprs.kallisto.counts, markers$`chemokine receptors`)

```

## Unbiased

Let's find unique positive markers for DR+27-. I filtered on FDR < 5% and FC > 1.5

```{r}
markers.unbiased <- data.table(limma.res.all)[adj.P.Val < .05 & logFC > log2(1.5), .N, by = .(feature)][N == 9, feature]
length(markers.unbiased)

anno.col.df <- metadata %>% 
  dplyr::filter(Position %in% colnames(exprs.kallisto.counts)) %>% 
  dplyr::mutate(cell = factor(cell, levels = cell.levels)) %>% 
  dplyr::arrange(cell)
row.names(anno.col.df) <- anno.col.df$Position


exprs.kallisto.counts[markers.unbiased, anno.col.df$Position, drop = F] %>% 
  pheatmap(labels_col = rep("", 90), 
           annotation_col = anno.col.df[, "cell", drop = F], 
           cluster_cols = F, cluster_rows = T, scale = "row")
```


```{r}
plot.markers(exprs.kallisto.counts, markers.unbiased)
```


# Pathways 

```{r}
library(gage)
do.gage <- function(top.table, genesets, rank.col, q.val.thresh = 1, gs.size.min = 10, gs.size.max = 500) {
  library(gage)
  library(magrittr)
  top.table %<>% dplyr::rename_("t" = rank.col)
  
  ### not sure if this is necessary but here for safety
  genesets <- genesets %>% lapply(function(gs.df) {
    intersect(gs.df, top.table$symbol)
  })  
  genesets <- genesets[which(sapply(genesets, length) >= gs.size.min & sapply(genesets, length) <= gs.size.max)] 
  print(sprintf("doing gage on %d sets", length(genesets)))
  
  gage.res <- data.frame(row.names = top.table$symbol, t = top.table$t) %>%
    gage(gsets = genesets, saaTest = gs.KSTest, set.size = c(gs.size.min, gs.size.max))
  gage.res <- data.frame(gage.res$greater) %>% tibble::rownames_to_column("geneset") %>% 
    dplyr::mutate(direction = "greater") %>%
    rbind(
      data.frame(gage.res$less) %>% tibble::rownames_to_column("geneset") %>% 
        dplyr::mutate(direction = "less")
    ) %>% 
    arrange(q.val, p.val) %>% 
    dplyr::select(geneset, p.val, q.val, set.size, direction) 
  data.table(gage.res)[, head(.SD[order(q.val, p.val)], 1), by = .(geneset)][
    q.val < q.val.thresh
    ][order(direction, q.val)]
}
```

## TF Analysis

TFs may be lowly and transiently expressed, so let's look at targets of TFs to infer TF activity in th cell. Note that ENCODE and Neph2012 encode targets with entrez IDs. For now, we won't use these. 

```{r}
library(tftargets)

tf.names <- do.call(rbind, list(data.frame(db = "ENCODE", tf = names(tftargets::ENCODE), stringsAsFactors = F),
                    data.frame(db = "ITFP", tf = names(tftargets::ITFP), stringsAsFactors = F),
                    data.frame(db = "Marbach2016", tf = names(tftargets::Marbach2016), stringsAsFactors = F),
                    data.frame(db = "TRED", tf = names(tftargets::TRED), stringsAsFactors = F),
                    data.frame(db = "Neph2012", tf = names(tftargets::Neph2012), stringsAsFactors = F),
                    data.frame(db = "TRRUST", tf = names(tftargets::TRRUST), stringsAsFactors = F)))  

# tftargets::ENCODE$ZBTB33
# tftargets::ITFP$AAAS
# tftargets::Marbach2016$AFP1
# tftargets::TRED$PAX1
# tftargets::Neph2012$`AG10803-DS12374`$AHR
# tftargets::TRRUST$AATF

```

Looking at all TRRUST TFs, four appear to have enriched targets in the DR+/27- vs all comparison. This also repeats for the memory and memoryEffector comparisons. 

```{r}
tfs <- intersect(markers$`transcription factors`, names(tftargets::TRRUST))

tf.enrich <- limma.res.all %>% 
  dplyr::filter(other.group == "all") %>% 
  dplyr::rename(symbol = feature) %>%
  do.gage(tftargets::TRRUST, "t", gs.size.min = 1)
subset(tf.enrich, q.val < .01)

```

Of these four, CIITA is especially overexpressed in DR+/27-. 

```{r}
tfs.potential <- subset(tf.enrich, q.val < .01)$geneset
plot.markers(exprs.kallisto.counts, tfs.potential)
```

The targets of CIITA are shown in the heatmap below. 

```{r}
anno.col.df <- metadata %>% 
  dplyr::filter(Position %in% colnames(exprs.kallisto.counts)) %>% 
  dplyr::mutate(cell = factor(cell, levels = cell.levels)) %>% 
  dplyr::arrange(cell)
row.names(anno.col.df) <- anno.col.df$Position

ciita.targets <- intersect(tftargets::TRRUST$CIITA, row.names(exprs.kallisto.counts))

exprs.kallisto.counts[ciita.targets, anno.col.df$Position] %>% 
  pheatmap(labels_col = rep("", 90), 
           annotation_col = anno.col.df[, "cell", drop = F], 
           cluster_cols = F, cluster_rows = T, scale = "row")
```





## DR+/27- vs groups

### C2: curated pathways

```{r, eval=FALSE, include=TRUE}
wb.gage <- createWorkbook()
gs <- MSigDB::MSigDB$C2_CURATED[grep("KEGG|PID|REACTOME|BIOCARTA", names(MSigDB::MSigDB$C2_CURATED))]
for (other.group in names(groups.to.compare)) {
  addWorksheet(wb.gage, other.group)
  
  ## Gage Pathway Enrichment 
  limma.res <- read.xlsx("../results/DiffExp.xlsx", sheet = other.group)
  gage.res <- limma.res %>% dplyr::rename(symbol = feature) %>% do.gage(gs, "t") %>%
    tidyr::separate(geneset, c("source", "pathway"), sep = "_", extra = "merge")
  
  writeDataTable(wb.gage, other.group, gage.res)

  saveWorkbook(wb.gage, file = "../results/gage_c2.xlsx", overwrite = TRUE)
}
```

### C7: immunological signatures

I found 30 signatures (paired: 15 up, 15 down) of genes in msigdb that are related either to Th1 vs Th17 experiments or CD4 vs NK experiments. 

```{r}
c7.names <- names(MSigDB::MSigDB$C7_IMMUNOLOGIC_SIGNATURES)
c7.th1.th17 <- c7.names[grep("Th1_vs_Th17", c7.names, ignore.case = T)]
c7.cd4.nk <- c7.names[grep("CD4.*_vs_NK", c7.names, ignore.case = T)]
gs.immuno <- MSigDB::MSigDB$C7_IMMUNOLOGIC_SIGNATURES[c(c7.th1.th17, c7.cd4.nk)]

wb.gage <- createWorkbook()
gage.res.c7 <- data.frame()
for (other.group in names(groups.to.compare)) {
  addWorksheet(wb.gage, other.group)
  limma.res <- read.xlsx("../results/DiffExp.xlsx", sheet = other.group)
  gage.res <- limma.res %>% dplyr::rename(symbol = feature) %>% do.gage(gs.immuno, "t") %>%
    tidyr::separate(geneset, c("source", "pathway"), sep = "_", extra = "merge") %>% 
    dplyr::arrange(q.val, direction)
  gage.res.c7 %<>% rbind(gage.res %>% dplyr::mutate(comparison = other.group))
  writeDataTable(wb.gage, other.group, gage.res)
}
saveWorkbook(wb.gage, file = "../results/gage_c7.xlsx", overwrite = TRUE)

```

As an example of how to interpret these results, let's look at DR+27- vs DR+27+. One study (GSE26030) with Th1 and Th17 cells provided significant enrichment. Genes associated with Th1 (vs Th17) cells 5 days post-polarization are upregulated in DR+27-. However, those associated with Th17 after 15 days are also upregulated in DR+27-. The top 10 upregulated genes from both genesets are shown below. The NK signature is more consistent. Genes associated with CD4 T cells (vs NK) are downregulated in DR+27-. Genes associated with NK cells (vs CD4 T Cells) are upregulated in DR+27-. 

```{r}
gage.res.c7 %>% 
  subset(comparison == "DRp27p") %>%
  dplyr::filter(q.val < .01) %>% 
  dplyr::select(source, pathway, q.val, direction)


limma.res.all %>% subset(other.group == "DRp27p") %>% 
  dplyr::filter(feature %in% gs.immuno[["GSE26030_TH1_VS_TH17_RESTIMULATED_DAY5_POST_POLARIZATION_UP"]]) %>% 
  dplyr::arrange(-t) %>% head(10)

limma.res.all %>% subset(other.group == "DRp27p") %>% 
  dplyr::filter(feature %in% gs.immuno[["GSE26030_TH1_VS_TH17_RESTIMULATED_DAY5_POST_POLARIZATION_DN"]]) %>% 
  dplyr::arrange(-t) %>% head(10)


```



<!-- ## Highly variable genes -->

<!-- We use TPM from kallisto to estimate 1000 highly variable genes. PCA on this is not as high quality (lower first eigenvalue, worse separation of classes) than with all the genes.  -->

<!-- ```{r} -->
<!-- gene.vars <- apply(exprs.kallisto.tpm, 1, sd) -->
<!-- vargenes <- gene.vars[order(-gene.vars)] %>% head(1000) %>% names -->
<!-- hist(gene.vars) -->

<!-- pca.res <- prcomp(log2(1 + exprs.kallisto.tpm[vargenes, samples.use])) -->
<!-- plot(pca.res$sdev/sum(pca.res$sdev)) -->
<!-- pca.res$rotation %>% data.frame %>%  -->
<!--   tibble::rownames_to_column("Position") %>% -->
<!--   dplyr::inner_join(metadata ) %>%  -->
<!--   ggplot(aes(PC1, PC2, col = cell, shape = status)) + geom_point(size = 3, alpha = .7) +  -->
<!--   scale_color_brewer(type = "qual") +  -->
<!--   geom_rangeframe() + theme_tufte() +  -->
<!--   labs(title = "Kallisto TPM") + theme(plot.title = element_text(hjust = 0.5, size = 16))   -->


<!-- ``` -->

























